Add support for convert/thumb options job format

felixbuenemann opened the pull request:
Dragonfly supports specifying the format and frame as a hash. This change
adds support for existing urls using this syntax for both convert and
thumb.

Example job created by dragonfly 1.0.7:

``` ruby
a.attachment.thumb('100x100', format: "png", frame: 1).to_a
=> [["f", "7a7/9fb/7a79fb0c87f4db253eb46aac26553351.jpg"],
    ["p", :thumb, "100x100", {:format=>"png", :frame=>1}]]
```

I added a check if `frame` is an integer, to protect against shell injection on unsigned urls.

I'm a bloody newb when it comes to Elixir, so please tell me if something should be done differently.

felixbuenemann commented:
This currently breaks when a sha secret is used, because `:crypto.hmac` fails to convert the format options struct to a binary via `:erlang.iolist_to_binary`. I'll check out how the original dragonfly gem handles this case.

felixbuenemann commented:
OK, added a `Steps.to_unique_string/1` to match the dragonfly gem's `to_unique_s` inclding sorting hash keys.

The Steps module is getting a bit large. Should I put the stringify code in a new `Stringifier` module?

cloud8421 commented:
Yes, definitely the string related logic could reside in its own module. It's also a perfect use case for a protocol (http://elixir-lang.org/getting-started/protocols.html), as in the code you're essentially doing a type check with a fallback. This kind of polymorphism is exactly what protocols were added for.

Let me know if you wanna give a refactor to protocols a stab or if you'd like me to do it and then review the resulting code.

Thanks!

felixbuenemann commented:
I'll look into protocols and give it a try. That would also mirror the ruby implementation which adds a `to_dragonfly_unique_s` method to core classes.

felixbuenemann commented:
@cloud8421 OK, I implemented the Steps.Stringify protocol. I'm not quite sure, wether this should be Steps.Stringify or Job.Stringify, what do you think?

cloud8421 commented:
Feels like it's a `Steps` concern to me.

felixbuenemann commented:
Should be good to merge then. Do you want me to squash first?

felixbuenemann committed:
Add support for convert/thumb options job format

Dragonfly supports specifying the format and frame as a hash. This change
adds support for existing urls using this syntax for both convert and
thumb.

Example job created by dragonfly 1.0.7:
```ruby
[["f", "7a7/9fb/7a79fb0c87f4db253eb46aac26553351.jpg"],
 ["p", :thumb, "100x100", {:format=>"png", :frame=>1}]]
```

This also adds a stringifier, which handles the creation of the unique
steps string that is used for signing the url. The previous
implementation could not handle structs and the required
canonicalization. This code mirrors the dragonfly gem's
to_unique_dragonfly_s core class extensions.

felixbuenemann commented:
Squashed and rebased.

cloud8421 commented:
:+1: Thanks!
